libgdf/src/join/full_join.cuh

```diff
@@ -0,0 +1,167 @@
+template <typename index_type, typename size_type>
+thrust::device_vector<index_type>
+create_missing_indices(
+        index_type const * const index_ptr,
+        const size_type max_index_value,
+		const size_type index_size) {
+	//Assume all the indices in invalid_index_map are invalid
+	thrust::device_vector<index_type> invalid_index_map(max_index_value, 1);
+	//Vector allocated for unmatched result
+	thrust::device_vector<index_type> unmatched_indices(max_index_value);
+	//Functor to check for index validity since left joins can create invalid indices
+	ValidRange<size_type> valid_range(0, max_index_value);
+
+	//invalid_index_map[index_ptr[i]] = 0 for i = 0 to max_index_value
+	//Thus specifying that those locations are valid
+	thrust::scatter_if(
+			thrust::device,
+			thrust::make_constant_iterator(0),
+			thrust::make_constant_iterator(0) + index_size,
+			index_ptr,//Index locations
+			index_ptr,//Stencil - Check if index location is valid
+			invalid_index_map.begin(),//Output indices
+			valid_range);//Stencil Predicate
+	size_type begin_counter = static_cast<size_type>(0);
+	size_type end_counter = static_cast<size_type>(invalid_index_map.size());
+	//Create list of indices that have been marked as invalid
+	size_type compacted_size = thrust::copy_if(
+			thrust::device,
+			thrust::make_counting_iterator(begin_counter),
+			thrust::make_counting_iterator(end_counter),
+			invalid_index_map.begin(),
+			unmatched_indices.begin(),
+			thrust::identity<index_type>()) -
+		unmatched_indices.begin();
+	unmatched_indices.resize(compacted_size);
+	return unmatched_indices;
+}
+
+/* --------------------------------------------------------------------------*/
+/**
+* @Synopsis  Expands a buffer's size
+*
+* @Param buffer Address of the buffer to expand
+* @Param buffer_capacity Memory allocated for buffer
+* @Param buffer_size Number of elements in the buffer
+* @Param expand_size Amount of extra elements to be pushed into the buffer
+* @tparam data_type The type of data associated with the buffer
+* @tparam size_type The data type used for size calculations
+*
+* @Returns  cudaSuccess upon successful completion of buffer expansion. Otherwise returns
+* the appropriate CUDA error code
+*/
+/* ----------------------------------------------------------------------------*/
+template <typename data_type, typename size_type>
+gdf_error expand_buffer(
+        data_type ** buffer,
+        size_type * const buffer_capacity,
+        const size_type buffer_size,
+        const size_type expand_size) {
+    size_type requested_size = buffer_size + expand_size;
+    //No need to proceed if the buffer can contain requested additional elements
+    if (*buffer_capacity >= requested_size) {
+        return GDF_SUCCESS;
+    }
+    data_type * new_buffer{nullptr};
+    data_type * old_buffer = *buffer;
+    CUDA_TRY( cudaMalloc(&new_buffer, requested_size*sizeof(data_type)) );
```

Comment:
This is a common functionality that will likely be used outside of the join algorithm, and therefore I think this entire function could be replaced by a generic `resize` function in RMM. It would take an existing allocation, a new size, and whatever else RMM would need to properly resize it. Thoughts @harrism ?

Reply:
It is slightly different than a resize. The buffer is only expanded if the requested size is greater than its current size.

Reply:
I believe RMM will track the size of the allocation? And can implement the same logic in which it only expands the allocation if necessary.

Reply:
rmmRealloc, although it is currently broken (doesn't do the copy!)